File: pylib/qlat/geometry.py
import cqlat as c

class Geometry:

    def __init__(self, total_site, multiplicity = None):
        if multiplicity is None:
            self.cdata = c.mk_geo(total_site)
        else:
            self.cdata = c.mk_geo(total_site, multiplicity)

    def __del__(self):
        c.free_geo(self)

    def __imatmul__(self, v1):
        assert isinstance(v1, Geometry)
        c.set_geo(self, v1)
        return self

    def copy(self, is_copying_data = True):
        x = Geometry([ 0, 0, 0, 0, ])
        if is_copying_data:
            x @= self
        return x

    def total_site(self):
        return c.get_total_site_geo(self)

    def total_volume(self):
        return c.get_total_volume_geo(self)

    def local_volume(self):
        return c.get_local_volume_geo(self)

    def multiplicity(self):
        return c.get_multiplicity_geo(self)

    def node_site(self):
        return c.get_node_site_geo(self)

    def eo(self):
        return c.get_eo_geo(self)

    def expansion_left(self):
        return c.get_expansion_left_geo(self)

    def expansion_right(self):
        return c.get_expansion_right_geo(self)

    def id_node(self):
        return c.get_id_node_geo(self)

    def num_node(self):
        return c.get_num_node_geo(self)

    def coor_node(self):
        return c.get_coor_node_geo(self)

    def size_node(self):
        return c.get_size_node_geo(self)

    def show(self):
        return "Geometry({}, {})".format(
                str(self.total_site()),
                self.multiplicity())

    def show_all(self):
        return "Geometry({}, {}, expansion_left={}, expansion_right={}, eo={})".format(
                str(self.total_site()),
                self.multiplicity(),
                str(self.expansion_left()),
                str(self.expansion_right()),
                self.eo())

    def coordinate_g_from_l(self, xl):
        return c.coordinate_g_from_l_geo(self, xl)

    def coordinate_l_from_g(self, xg):
        return c.coordinate_l_from_g_geo(self, xg)

    def is_local(self, xl):
        return c.is_local_geo(self, xl)

###

def geo_reform(geo, multiplicity = 1, expansion_left = None, expansion_right = None):
    assert isinstance(geo, Geometry)
    if expansion_left is None:
        expansion_left = [ 0, 0, 0, 0, ]
    elif isinstance(expansion_left, int):
        e = expansion_left
        expansion_left = [ e, e, e, e, ]
    if expansion_right is None:
        expansion_right = [ 0, 0, 0, 0, ]
    elif isinstance(expansion_right, int):
        e = expansion_right
        expansion_right = [ e, e, e, e, ]
    geo_new = Geometry([ 0, 0, 0, 0, ])
    c.set_geo_reform(geo_new, geo, multiplicity, expansion_left, expansion_right)
    return geo_new

def geo_eo(geo, eo = 0):
    assert isinstance(geo, Geometry)
    geo_new = Geometry([ 0, 0, 0, 0, ])
    c.set_geo_eo(geo_new, geo, eo)
    return geo_new

File: qlat/pylib/cqlat/geometry.cpp
#include "lib.h"

EXPORT(mk_geo, {
  using namespace qlat;
  Geometry* pgeo = new Geometry();
  return py_convert((void*)pgeo);
})

EXPORT(free_geo, {
  using namespace qlat;
  return free_obj<Geometry>(args);
})

EXPORT(set_geo, {
  using namespace qlat;
  return set_obj<Geometry>(args);
})

EXPORT(set_geo_total_site, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  PyObject* p_total_site = NULL;
  int multiplicity = 1;
  if (!PyArg_ParseTuple(args, "OO|i", &p_geo, &p_total_site, &multiplicity)) {
    return NULL;
  }
  Geometry& geo = py_convert_type<Geometry>(p_geo);
  const Coordinate total_site = py_convert_data<Coordinate>(p_total_site);
  geo.init(total_site, multiplicity);
  Py_RETURN_NONE;
})

EXPORT(set_geo_reform, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  int multiplicity = 1;
  PyObject* p_expansion_left = NULL;
  PyObject* p_expansion_right = NULL;
  if (!PyArg_ParseTuple(args, "O|iOO", &p_geo, &multiplicity, &p_expansion_left,
                        &p_expansion_right)) {
    return NULL;
  }
  Geometry& geo = py_convert_type<Geometry>(p_geo);
  Coordinate expansion_left, expansion_right;
  if (NULL != p_expansion_left) {
    py_convert(expansion_left, p_expansion_left);
  }
  if (NULL != p_expansion_right) {
    py_convert(expansion_right, p_expansion_right);
  }
  geo.remult(multiplicity);
  geo.resize(expansion_left, expansion_right);
  Py_RETURN_NONE;
})

EXPORT(set_geo_eo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  int eo = 0;
  if (!PyArg_ParseTuple(args, "O|i", &p_geo, &eo)) {
    return NULL;
  }
  Geometry& geo = py_convert_type<Geometry>(p_geo);
  geo.eo = eo;
  Py_RETURN_NONE;
})

EXPORT(get_total_site_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.total_site());
})

EXPORT(get_local_volume_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.local_volume());
})

EXPORT(get_total_volume_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.total_volume());
})

EXPORT(get_multiplicity_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.multiplicity);
})

EXPORT(get_node_site_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.node_site);
})

EXPORT(get_eo_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.eo);
})

EXPORT(get_expansion_left_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.expansion_left);
})

EXPORT(get_expansion_right_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.expansion_right);
})

EXPORT(get_id_node_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.geon.id_node);
})

EXPORT(get_num_node_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.geon.num_node);
})

EXPORT(get_coor_node_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.geon.coor_node);
})

EXPORT(get_size_node_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  return py_convert(geo.geon.size_node);
})

EXPORT(coordinate_g_from_l_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  PyObject* p_xl = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_geo, &p_xl)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  Coordinate xl;
  py_convert(xl, p_xl);
  return py_convert(geo.coordinate_g_from_l(xl));
})

EXPORT(coordinate_l_from_g_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  PyObject* p_xg = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_geo, &p_xg)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  Coordinate xg;
  py_convert(xg, p_xg);
  return py_convert(geo.coordinate_l_from_g(xg));
})

EXPORT(is_local_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  PyObject* p_xl = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_geo, &p_xl)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  const Coordinate xl = py_convert_data<Coordinate>(p_xl);
  return py_convert(geo.is_local(xl));
})

EXPORT(is_local_xg_geo, {
  using namespace qlat;
  PyObject* p_geo = NULL;
  PyObject* p_xg = NULL;
  if (!PyArg_ParseTuple(args, "OO", &p_geo, &p_xg)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  const Coordinate xg = py_convert_data<Coordinate>(p_xg);
  const Coordinate xl = geo.coordinate_l_from_g(xg);
  return py_convert(geo.is_local(xl));
})

EXPORT(get_xg_list, {
  // return xg for all local sites
  using namespace qlat;
  PyObject* p_geo = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_geo)) {
    return NULL;
  }
  const Geometry& geo = py_convert_type<Geometry>(p_geo);
  vector<Coordinate> xgs(geo.local_volume());
  qthread_for(index, geo.local_volume(), {
    const Coordinate xl = geo.coordinate_from_index(index);
    const Coordinate xg = geo.coordinate_g_from_l(xl);
    xgs[index] = xg;
  })
  return py_convert(get_data(xgs));
})
